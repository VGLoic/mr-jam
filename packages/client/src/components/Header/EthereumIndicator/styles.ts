import { makeStyles, Theme } from "@material-ui/core";

const styles = ({ spacing, palette, transitions }: Theme): Styles => ({
    notFound: {
        color: palette.error.main,
    },
    notEnabled: {
        animation: `$alternate-red 2000ms ${transitions.easing.easeInOut} 0ms infinite`
    },
    wrongNetwork: {
        color: palette.warning.main
    },
    correctNetwork: {
        color: palette.success.main
    },
    aloneIcon: {
        margin: `0 ${spacing(1) + spacing(1) / 2}px`
    },
    "@keyframes alternate-red": {
        "0%": {
            color: "default"
        },
        "50%": {
            color: palette.error.main
        },
        "100%": {
            color: "default"
        }
    }
});

export interface Styles {
    notFound: any;
    wrongNetwork: any;
    correctNetwork: any;
    aloneIcon: any;
    "@keyframes alternate-red": any,
    notEnabled: any;
}

export const useStyles = makeStyles(styles);
