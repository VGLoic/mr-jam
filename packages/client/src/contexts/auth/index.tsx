import React, {
  createContext,
  useState,
  useContext,
  useMemo,
  useEffect
} from "react";
import { useLocation, useHistory } from "react-router-dom";
// Utils
import AuthUtils, { LoginInformations } from "utils/auth.utils";

interface IContext {
  isAuthenticated: boolean;
  isAuthInitialized: boolean;
}

const AuthContext: React.Context<any> = createContext(undefined);

type AuthProviderProps = any;
export const AuthProvider = (props: AuthProviderProps) => {
  const [isAuthenticated, setIsAuthenticated] = useState<boolean>(
    AuthUtils.isAuthenticated()
  );

  const { pathname, hash } = useLocation();
  const history = useHistory();
  const [isAuthInitialized, setAuthInitialized] = useState<boolean>(false);

  useEffect((): void => {
    if (!isAuthenticated) {
      const {
        token,
        expirationDate,
        originalUrl
      }: LoginInformations = AuthUtils.deriveLoginInfoFromUrl(pathname, hash);
      // Authenticating
      if (token && expirationDate) {
        AuthUtils.setToken(token);
        AuthUtils.setExpirationDate(expirationDate);
        setIsAuthenticated(true);
        history.push(originalUrl || "/");
      } else {
        AuthUtils.clearAuthData();
        AuthUtils.redirectToLogin(pathname);
      }
    }
    setAuthInitialized(true);
  }, [pathname, hash, history, isAuthenticated]);

  const value = useMemo(
    () => ({
      isAuthenticated,
      isAuthInitialized
    }),
    [isAuthenticated, isAuthInitialized]
  );

  return <AuthContext.Provider value={value} {...props} />;
};

export interface IAuth {
  isAuthInitialized: boolean;
  isAuthenticated: boolean;
  getToken: () => string | null;
}

export const useAuth = (): IAuth => {
  const context: IContext = useContext(AuthContext);

  if (!context) {
    throw new Error(
      "useAuth must be accessed in a child of the AuthContextProvider."
    );
  }

  return {
    isAuthInitialized: context.isAuthInitialized,
    isAuthenticated: context.isAuthenticated,
    getToken: AuthUtils.getToken
  };
};
