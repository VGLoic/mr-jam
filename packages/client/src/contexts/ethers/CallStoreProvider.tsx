import React, { createContext, useReducer, Context } from "react";

export type Key = string;

export interface CallObject {
  key: Key;
  loading: boolean;
  error: string | null;
  data: any | null;
}
export type CallStore = Record<Key, CallObject>;

export const LOADING = "LOADING";
interface LoadingAction {
  type: typeof LOADING;
  payload: Key;
}
export const SUCCESS = "SUCCESS";
export interface CallPayload {
  key: Key;
  data: any;
}
interface SuccessAction {
  type: typeof SUCCESS;
  payload: CallPayload;
}
export const ERROR = "ERROR";
export interface ErrorPayload {
  key: Key;
  error: string;
}
interface ErrorAction {
  type: typeof ERROR;
  payload: ErrorPayload;
}

type Action = LoadingAction | SuccessAction | ErrorAction;
const initialState: CallStore = {};

const reducer = (state: CallStore, action: Action): CallStore => {
  let callObject: CallObject | undefined;
  switch (action.type) {
    case LOADING:
      callObject = state[action.payload];
      if (callObject) {
        return {
          ...state,
          [action.payload]: {
            ...callObject,
            loading: true,
          },
        };
      }
      return {
        ...state,
        [action.payload]: {
          key: action.payload,
          loading: true,
          error: null,
          data: null,
        },
      };
    case SUCCESS:
      callObject = state[action.payload.key];
      if (!callObject) return state;
      return {
        ...state,
        [action.payload.key]: {
          ...callObject,
          loading: false,
          data: action.payload.data,
        },
      };
    case ERROR:
      callObject = state[action.payload.key];
      if (!callObject) return state;
      return {
        ...state,
        [action.payload.key]: {
          ...callObject,
          loading: false,
          error: action.payload.error,
        },
      };
    default:
      throw new Error("Unknown action type");
  }
};

export interface ICallStoreContext {
  callStore: CallStore;
  dispatch: React.Dispatch<Action>;
}

export const CallContext: Context<
  ICallStoreContext | undefined
> = createContext(undefined as ICallStoreContext | undefined);

type CallStoreProviderProps = any;
const CallStoreProvider = (props: CallStoreProviderProps) => {
  const [state, dispatch] = useReducer(reducer, initialState);

  const value: ICallStoreContext = {
    callStore: state,
    dispatch,
  };

  return <CallContext.Provider value={value} {...props} />;
};

export default CallStoreProvider;
