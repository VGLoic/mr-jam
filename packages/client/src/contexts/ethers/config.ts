import ProjectBuild from "smart-contracts/build/contracts/Project.json";
import ProjectRegistryBuild from "smart-contracts/build/contracts/ProjectRegistry.json";

export const allowedNetworks: Record<number, string> = {
    5777: "development"
};

export interface ContractMetadata {
    abi: any[];
    bytecode: string;
    networks?: Record<number, string>
}

export enum Contracts {
    ProjectRegistry = "ProjectRegistry",
    Project = "Project"
}

export interface ContractMetadatas {
    [Contracts.Project]: ContractMetadata;
    [Contracts.ProjectRegistry]: ContractMetadata
}

export const contractMetadatas: ContractMetadatas = {
    [Contracts.Project]: {
        abi: ProjectBuild.abi,
        bytecode: ProjectBuild.bytecode
    },
    [Contracts.ProjectRegistry]: {
        abi: ProjectRegistryBuild.abi,
        bytecode: ProjectRegistryBuild.bytecode,
        networks: Object.entries(ProjectRegistryBuild.networks).reduce(
            (acc: Record<number, string>, [networkId, networkInfo]: [string, any]) => ({
                ...acc,
                [parseInt(networkId)]: networkInfo.address
            }),
            {}
        )
    }
};
