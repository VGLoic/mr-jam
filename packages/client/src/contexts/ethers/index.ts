export * from "./EthersProvider";
export * from "./TransactionStoreProvider";
