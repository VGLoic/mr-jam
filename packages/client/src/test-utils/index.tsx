import React from "react";
import { render } from '@testing-library/react'
import MockedProvider from "./MockedProvider";

const customRender = (ui: React.ReactElement, options?: any) => {
  const RenderedMockedProvider = (props: any) => <MockedProvider {...props} ethersContext={options?.ethersContext} />;
  return render(ui, { wrapper: RenderedMockedProvider, ...options });
}

// re-export everything
export * from '@testing-library/react'

// override render method
export { customRender as render }