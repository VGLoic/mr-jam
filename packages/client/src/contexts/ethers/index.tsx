import React, { createContext, Context, useContext, useState, useEffect } from "react";
import { ethers,  } from "ethers";

export interface IEthereum {
    isMetaMask: boolean;
    networkVersion: string;
    selectedAddress: string | null;
    enable: () => Promise<void>;
    on: (eventName: string, callback: (args: any) => any) => void;
    removeListener: (eventName: string, callback: (args: any) => any) => void;
}

type Ethereum = IEthereum | null;

export interface IEthersContext {
    ethereum: Ethereum;
    isMetaMaskDetected: boolean;
    isEnabled: boolean;
    setIsEnabled: React.Dispatch<React.SetStateAction<boolean>>;
    isNetworkAllowed: boolean;
    provider: ethers.providers.Web3Provider | null
}

const EthersContext: Context<IEthersContext | undefined> = createContext(undefined as IEthersContext | undefined);

type EthersProviderProps = any;

export const allowedNetworks: Record<number, string> = {
    5777: "development"
};

export const METAMASK_ENABLED_KEY: string = "mr-explorer_metamask_enabled";
export const METAMASK_ENABLED_VALUE: string = "yeay";

export const EthersProvider = (props: EthersProviderProps) => {
    const ethereum: Ethereum = (window as any).ethereum || null;

    const isMetaMaskDetected = Boolean(ethereum?.isMetaMask);

    const [isEnabled, setIsEnabled] = useState<boolean>(
        localStorage.getItem(METAMASK_ENABLED_KEY) === METAMASK_ENABLED_VALUE
    );

    const [isNetworkAllowed, setIsNetworkAllowed] = useState<boolean>(
        Boolean(
            ethereum?.networkVersion &&
            ethereum.networkVersion in allowedNetworks
        )
    );

    useEffect((): () => void => {
        const onAccountsChanged = (accounts: string[]): void => {
            if (accounts.length === 0) {
                localStorage.removeItem(METAMASK_ENABLED_KEY);
                setIsEnabled(false);
            }
        }
        const onNetworkChanged = (networkId: number): void => {
            setIsNetworkAllowed(networkId in allowedNetworks);
        };
        ethereum?.on("accountsChanged", onAccountsChanged);
        ethereum?.on("networkChanged", onNetworkChanged);
        return () => {
            ethereum?.removeListener("accountsChanged", onAccountsChanged);
            ethereum?.removeListener("networkChanged", onNetworkChanged);
        }
    }, [ethereum]);

    const provider = ethereum ? new ethers.providers.Web3Provider(ethereum) : null;

    const value: IEthersContext = {
        ethereum,
        isMetaMaskDetected,
        isEnabled,
        setIsEnabled,
        isNetworkAllowed,
        provider
    };

    return <EthersContext.Provider value={value} {...props} />
}

export interface UseEthers {
    isMetaMaskDetected: boolean;
    isEnabled: boolean;
    enableMetaMask: () => Promise<void>;
    isNetworkAllowed: boolean;
    provider: ethers.providers.Web3Provider | null;
}

export const useEthers = (): UseEthers => {
    const context: IEthersContext | undefined = useContext(EthersContext);

    if (!context) {
        throw new Error(
          "useEthers must be accessed in a child of the EthersProvider."
        );
    }

    const enableMetaMask = async (): Promise<void> => {
        try {
            await context.ethereum?.enable();
            context.setIsEnabled(true);
            localStorage.setItem(METAMASK_ENABLED_KEY, METAMASK_ENABLED_VALUE);
        } catch (err) {
            console.error("User has canceled the connection request. MetaMask error: ", err);
            localStorage.removeItem(METAMASK_ENABLED_KEY);
        }
    }

    return {
        isMetaMaskDetected: context.isMetaMaskDetected,
        isEnabled: context.isEnabled,
        enableMetaMask,
        isNetworkAllowed: context.isNetworkAllowed,
        provider: context.provider
    };
}