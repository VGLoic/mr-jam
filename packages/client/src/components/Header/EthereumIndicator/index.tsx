import React from "react";
import classnames from "classnames";
// UI Components
import { Tooltip, IconButton, SvgIcon } from "@material-ui/core";
import { mdiEthereum } from "@mdi/js";
// Hooks
import { useEthers } from "contexts/ethers";
import { allowedNetworks } from "contexts/ethers/config";
// Styles
import { useStyles } from "./styles";

const networkString = Object.entries(allowedNetworks).reduce(
  (acc: string, [networkId, networkName]): string => {
    const networkAddition = `${networkName} (ID ${networkId})`;
    if (acc) {
      return `${acc}, ${networkAddition}`;
    }
    return networkAddition;
  },
  ""
);

const EthereumIndicator = () => {
  const {
    isMetaMaskDetected,
    isEnabled,
    enableMetaMask,
    isNetworkAllowed,
  } = useEthers();

  const classes = useStyles();

  if (!isMetaMaskDetected) {
    return (
      <Tooltip title="MetaMask extension is needed in order to use blockchain features.">
        <SvgIcon
          aria-label="MetaMask extension is needed in order to use blockchain features."
          className={classnames(classes.notFound, classes.aloneIcon)}
        >
          <path d={mdiEthereum} />
        </SvgIcon>
      </Tooltip>
    );
  }

  if (!isEnabled) {
    return (
      <Tooltip title="Enable MetaMask">
        <IconButton
          data-testid="enable-metamask-button"
          aria-label="Enable MetaMask"
          onClick={enableMetaMask}
        >
          <SvgIcon className={classes.notEnabled}>
            <path d={mdiEthereum} />
          </SvgIcon>
        </IconButton>
      </Tooltip>
    );
  }

  if (!isNetworkAllowed) {
    return (
      <Tooltip
        title={`Please, change on one of the allowed following networks ${networkString}.`}
      >
        <SvgIcon
          aria-label={`Please, change on one of the following networks ${networkString}.`}
          className={classnames(classes.wrongNetwork, classes.aloneIcon)}
        >
          <path d={mdiEthereum} />
        </SvgIcon>
      </Tooltip>
    );
  }

  return (
    <Tooltip title="MetaMask is properly setup">
      <SvgIcon
        aria-label="MetaMask is properly setup"
        className={classnames(classes.correctNetwork, classes.aloneIcon)}
      >
        <path d={mdiEthereum} />
      </SvgIcon>
    </Tooltip>
  );
};

export default EthereumIndicator;
