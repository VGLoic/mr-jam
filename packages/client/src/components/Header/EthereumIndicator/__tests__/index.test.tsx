import React from "react";
import { render } from "test-utils";
import { ethers } from "ethers";

import EthereumIndicator from "../index";
import { UseEthers, useEthers, EthersProvider } from "contexts/ethers";

jest.mock("contexts/ethers");

describe("<EthereumIndicator />", () => {
  const baseMock: UseEthers = {
    isMetaMaskDetected: true,
    isEnabled: true,
    isNetworkAllowed: true,
    enableMetaMask: jest.fn(() => Promise.resolve()),
    provider: {} as ethers.providers.Web3Provider,
  };
  beforeAll(() => {
    (EthersProvider as jest.Mock<
      JSX.Element
    >).mockImplementation(({ children }) => <div>{children}</div>);
  });
  test("when MetaMask is not detected", () => {
    (useEthers as jest.Mock<UseEthers>).mockReturnValue({
      ...baseMock,
      isMetaMaskDetected: false,
    });
    const { asFragment } = render(<EthereumIndicator />);
    expect(asFragment()).toMatchSnapshot("MetaMask not detected");
  });
  test("when MetaMask is not enabled", () => {
    (useEthers as jest.Mock<UseEthers>).mockReturnValue({
      ...baseMock,
      isEnabled: false,
    });
    const { asFragment } = render(<EthereumIndicator />);
    expect(asFragment()).toMatchSnapshot("MetaMask not enabled");
  });

  test("when network is not allowed", () => {
    (useEthers as jest.Mock<UseEthers>).mockReturnValue({
      ...baseMock,
      isNetworkAllowed: false,
    });
    const { asFragment } = render(<EthereumIndicator />);
    expect(asFragment()).toMatchSnapshot("Network not allowed");
  });
});
