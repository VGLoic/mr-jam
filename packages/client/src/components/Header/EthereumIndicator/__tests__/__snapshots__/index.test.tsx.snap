// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<EthereumIndicator /> when MetaMask is not detected: MetaMask not detected 1`] = `
<DocumentFragment>
  <div>
    <svg
      aria-hidden="true"
      aria-label="MetaMask extension is needed in order to use blockchain features."
      class="MuiSvgIcon-root makeStyles-notFound-1 makeStyles-aloneIcon-5"
      focusable="false"
      title="MetaMask extension is needed in order to use blockchain features."
      viewBox="0 0 24 24"
    >
      <path
        d="M12,1.75L5.75,12.25L12,16L18.25,12.25L12,1.75M5.75,13.5L12,22.25L18.25,13.5L12,17.25L5.75,13.5Z"
      />
    </svg>
  </div>
</DocumentFragment>
`;

exports[`<EthereumIndicator /> when MetaMask is not enabled: MetaMask not enabled 1`] = `
<DocumentFragment>
  <div>
    <button
      aria-label="Enable MetaMask"
      class="MuiButtonBase-root MuiIconButton-root"
      data-testid="enable-metamask-button"
      tabindex="0"
      title="Enable MetaMask"
      type="button"
    >
      <span
        class="MuiIconButton-label"
      >
        <svg
          aria-hidden="true"
          class="MuiSvgIcon-root makeStyles-notEnabled-28"
          focusable="false"
          viewBox="0 0 24 24"
        >
          <path
            d="M12,1.75L5.75,12.25L12,16L18.25,12.25L12,1.75M5.75,13.5L12,22.25L18.25,13.5L12,17.25L5.75,13.5Z"
          />
        </svg>
      </span>
      <span
        class="MuiTouchRipple-root"
      />
    </button>
  </div>
</DocumentFragment>
`;

exports[`<EthereumIndicator /> when network is not allowed: Network not allowed 1`] = `
<DocumentFragment>
  <div>
    <svg
      aria-hidden="true"
      aria-label="Please, change on one of the following networks development (ID 5777)."
      class="MuiSvgIcon-root makeStyles-wrongNetwork-77 makeStyles-aloneIcon-79"
      focusable="false"
      title="Please, change on one of the allowed following networks development (ID 5777)."
      viewBox="0 0 24 24"
    >
      <path
        d="M12,1.75L5.75,12.25L12,16L18.25,12.25L12,1.75M5.75,13.5L12,22.25L18.25,13.5L12,17.25L5.75,13.5Z"
      />
    </svg>
  </div>
</DocumentFragment>
`;
